<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Thomas Ng and Eric C. Anderson" />

<meta name="date" content="2019-09-27" />

<title>An Introduction to microhaplot data preparation</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' || rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>

</head>

<body>




<h1 class="title toc-ignore">An Introduction to microhaplot data preparation</h1>
<h4 class="author"><em>Thomas Ng and Eric C. Anderson</em></h4>
<h4 class="date"><em>2019-09-27</em></h4>



<p>This vignette concerns itself with an overview of microhaplot and on preparation of data for input into microhaplot Once your data has been transferred to the appropriate Haplot folder, see the <code>microhaplot walkthrough</code> vignette to learn more about microhaplot’s graphical features.</p>
<div id="how-does-microhaplot-work" class="section level2">
<h2>How does microhaplot work?</h2>
<p><code>microhaplot</code> is a tool that makes it easy to extract, visualize, and interpret short-read haplotypes—or “microhaplotypes” in the parlance of some—from aligned sequences. Before we go further, it will be good to step back and confirm of what we are calling as a short-read haplotype, or a “microhaplotype” within the context of microhaplot.</p>
<blockquote>
<p><em>What is a microhaplotype?</em></p>
</blockquote>
<p>To define a microhaplotype you need to first define a reference point or position that a read might align to. Then, <code>microhaplot</code> can analyze the reads aligning at those positions, and extracts the variant sequence of the reads at those positions. For example, let’s say that we know SNPs occur at positions 19, 27, 68, and 101 of a particular reference sequence that we have aligned our reads to. If a read aligns to the ref and shows an <code>A</code> at position 19, a <code>G</code> at position 27, a <code>C</code> at position 68, and another <code>C</code> at position 101, then we say the microhaplotype on that read is <code>AGCC</code>.</p>
<p><code>microhaplot</code> is a tool that</p>
<ol style="list-style-type: decimal">
<li>automates the extraction of these haplotypes from aligned data (typically amplicon sequencing data)</li>
<li>provides useful filtering and visualizing capabilities for understanding your microhaplotype data and for calling genotypes at microhaplotypes.</li>
</ol>
<p>To understand what is involved in getting your data into <code>microhaplot</code>, it will be useful to have a brief introduction to what it does and how it operates. We will start by listing what it is <strong>not</strong>.</p>
<ul>
<li><code>microhaplot</code> <strong>is not a SNP-caller or variant detector</strong>. It is not its job to determine if a variant might exist in a particular location or not. We assume you already know what variant position you want to look for by supplying <code>microhaplot</code> a variant caller file (VCF). <code>microhaplot</code> just extracts the bases sequenced on different reads at those positions.</li>
<li><code>microhaplot</code> <strong>is not an alignment program</strong>. <code>microhaplot</code> does not do alignments. You have to provide aligned data to it. From the aligned data, <code>microhaplot</code> can extract the bases sequenced at the desired locations.</li>
<li><code>microhaplot</code> <strong>is not a haplotype phasing program</strong>. There is no special algorithm going on in <code>microhaplot</code> to try to assemble haplotypes (across reads or loci, for example.) It merely uses the simple fact that a single sequencing read will come from a single chromosome. So, if you can draw bases present at 4 variant positions along a 150-bp read, you know that those 4 bases constitute a haplotype.<br />
</li>
<li><code>microhaplot</code> <strong>is not a de-novo assembly algorithm</strong>. This package will not infer or sort out any physical linkage between reads of different reference loci.</li>
</ul>
</div>
<div id="what-kind-of-data-can-i-use-with-microhaplot" class="section level2">
<h2>What kind of data can I use with <code>microhaplot</code>?</h2>
<p>‘microhaplot’ is designed to work with short read sequencing data where there is a relatively small number sequencing start sites. It is emphatically <em>not</em> geared to whole-genome shotgun sequencing data in which you may have overlapping reads with different starting place. We use ‘microhaplot’ primarily with amplicon sequencing data in which a few hundred fragments (of about 100 - 150 bp each) are amplified by PCR in a few hundred individuals. The DNA from different individuals is bar-coded so that it may be demultiplexed later, and then these amplified fragments are sequenced (usually to fairly high read depth) on an Illumina platform.</p>
<p>Though we use ‘microhaplot’ for amplicon sequencing, it might also be useful for analyzing haplotype data from experiments that sequence captured or baited fragments, (e.g., baited DD-Rad, RAPTURE, myBaits, ExonCapture, etc.), and might even be applicable to standard ddRAD or RAD-PE experiments, however, as the number of loci increases and average read depths per individual-locus combination decreases, the unevenness of coverage can become a bigger problem, (and we have more reservations about allelic dropout/null alleles when using digestion-based protocols (ddRAD, Rapture, etc.) than we do with amplicon sequencing).</p>
</div>
<div id="why-did-we-write-microhaplot" class="section level2">
<h2>Why did we write ‘microhaplot’?</h2>
<p>We tried to use existing tools to give us what we needed for our amplicon-based microhaplotypes.</p>
<ul>
<li>We were hopeful that <a href="http://catchenlab.life.illinois.edu/stacks/">STACKS</a> would provide us useful microhaplotypes. What we found was that the presence of indels caused STACKS to fail, and even if it didn’t we never found a reliable way to export inferred read haplotypes. (We could see what looked like haplotypes in the SQL-based viewer, but couldn’t figure out how to get at those in an exported fashion!)</li>
<li>We spent countless months trying to get <a href="https://github.com/ekg/freebayes">freebayes</a> to properly call microhaplotypes. Despite being an excellent variant detector, however, when presented with situations that were clearly simple microhaplotypes, <code>freebayes</code> would fail to call them, or would return results in a very difficult to manage “multinucleotide polymorphism” in the VCF.<br />
</li>
<li>We thought that <code>GATK's Unified Haplotype Caller</code> might deliver us from the bioinformatic hell we were experiencing, but alas, with the super-high read depths at some of the fragments in our amplicon sequencing data, GATK was brought to its knees and didn’t offer a viable solution. It appears that the HaplotypeCaller is trying to solve a much harder problem than just extracting short-read haplotypes.</li>
</ul>
<p>After a few months of this, we realized that there is no software tailored for the relatively straightforward task of extracting and analyzing microhaplotypes. The other software programs have all been designed for different purposes, and, as a consequence, while they work great for identifying variants or performing de-novo assemblies, one might not expect them to work for the specialized task of extracting and analyzing microhaplotypes.</p>
</div>
<div id="how-specifically-should-i-prepare-my-data" class="section level2">
<h2>How, specifically, should I prepare my data?</h2>
<p>The starting point for <code>microhaplot</code> requires:</p>
<ol style="list-style-type: decimal">
<li>An indexed, SAM file of reads aligned to the reference sequences for each amplicon.</li>
<li>A VCF file whose contents define the positions in each reference sequence from which you want to extract variants into microhaplotypes. Note that the <em>individuals</em> that appear in the VCF file need not be the same ones whose reads appear in the assemblies. The VCF is merely used as a way of defining positions to extract. Still, since the content of microhaplotype is based on the variant positions listed in the VCF file, it is crucial for the VCF file to only contain highly reliable variant sites.</li>
</ol>
<p>We describe here the workflow that we use to get these two necessary files from the short-read amplicon sequencing data that come off our Illumina Mi-seq. This workflow can be tweaked and tailored for your own data. Each step is described in one of the following subsections.</p>
<div id="starting-state" class="section level3">
<h3>Starting state</h3>
<p>Out description starts at the point where we have copied all of our fastq.gz files from the sequencer into a directory called <code>rawdata</code>. Next to that directory we have made two more directories, <code>flash</code> and <code>map</code>, in which we will be creating files and doing work. We have named all the Illumina files so that a directory listing of <code>rawdata</code> looks like this:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="ex">/rawdata/--%</span> ll <span class="kw">|</span> <span class="fu">head</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="ex">total</span> 3517864</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="ex">drwxrwxr-x</span> 2 biopipe biopipe  122880 Aug 18 11:57 ./</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="ex">drwxrwxr-x</span> 5 biopipe biopipe    4096 Aug 18 11:31 ../</a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="ex">-rw-r--r--</span> 1 biopipe biopipe  672293 Aug 18 11:53 satro_S100_L001_I1_001.fastq.gz</a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="ex">-rw-r--r--</span> 1 biopipe biopipe  725385 Aug 18 11:53 satro_S100_L001_I2_001.fastq.gz</a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="ex">-rw-r--r--</span> 1 biopipe biopipe 4656145 Aug 18 11:53 satro_S100_L001_R1_001.fastq.gz</a>
<a class="sourceLine" id="cb1-8" data-line-number="8"><span class="ex">-rw-r--r--</span> 1 biopipe biopipe 4749637 Aug 18 11:53 satro_S100_L001_R2_001.fastq.gz</a>
<a class="sourceLine" id="cb1-9" data-line-number="9"><span class="ex">-rw-r--r--</span> 1 biopipe biopipe  635863 Aug 18 11:53 satro_S101_L001_I1_001.fastq.gz</a>
<a class="sourceLine" id="cb1-10" data-line-number="10"><span class="ex">-rw-r--r--</span> 1 biopipe biopipe  711534 Aug 18 11:53 satro_S101_L001_I2_001.fastq.gz</a>
<a class="sourceLine" id="cb1-11" data-line-number="11"><span class="ex">-rw-r--r--</span> 1 biopipe biopipe 4410826 Aug 18 11:53 satro_S101_L001_R1_001.fastq.gz</a></code></pre></div>
<p>In this case, there are 384 such sets of files. The number after the capital S in the filename gives the ID number of each individual fish.</p>
</div>
<div id="flashing-the-reads-together" class="section level3">
<h3>Flashing the reads together</h3>
<p>Our data are paired end reads of fragments that are short enough that the primary reads and the paired-end reads overlap. Rather than treat them as two separate reads, we combine them together into a single read using <a href="https://ccb.jhu.edu/software/FLASH/">flash</a>. Here is the command we use to cycle over all the files and flash them, run from the <code>flash</code> directory which is at the same level as the <code>rawdata</code> directory. (Note that the <code>/flash/--%</code> in the following code fragment is just the command prompt—it tells you which directory we are in…)</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="ex">/flash/--%</span> for i in <span class="dt">{1..384}</span><span class="kw">;</span> <span class="kw">do</span> <span class="ex">flash</span> -m 10 -M 100 -z --output-prefix=S<span class="va">${i}</span> ../rawdata/satro_S<span class="va">${i}</span>_L001_R1_001.fastq.gz ../rawdata/satro_S<span class="va">${i}</span>_L001_R2_001.fastq.gz<span class="kw">;</span> <span class="kw">done</span></a></code></pre></div>
<p>The output that <code>flash</code> spits out to stdout looks like this:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb3-1" data-line-number="1">[<span class="ex">FLASH</span>] Starting FLASH v1.2.11</a>
<a class="sourceLine" id="cb3-2" data-line-number="2">[<span class="ex">FLASH</span>] Fast Length Adjustment of SHort reads</a>
<a class="sourceLine" id="cb3-3" data-line-number="3">[<span class="ex">FLASH</span>]</a>
<a class="sourceLine" id="cb3-4" data-line-number="4">[<span class="ex">FLASH</span>] Input files:</a>
<a class="sourceLine" id="cb3-5" data-line-number="5">[<span class="ex">FLASH</span>]     ../rawdata/satro_S1_L001_R1_001.fastq.gz</a>
<a class="sourceLine" id="cb3-6" data-line-number="6">[<span class="ex">FLASH</span>]     ../rawdata/satro_S1_L001_R2_001.fastq.gz</a>
<a class="sourceLine" id="cb3-7" data-line-number="7">[<span class="ex">FLASH</span>]</a>
<a class="sourceLine" id="cb3-8" data-line-number="8">[<span class="ex">FLASH</span>] Output files:</a>
<a class="sourceLine" id="cb3-9" data-line-number="9">[<span class="ex">FLASH</span>]     ./S1.extendedFrags.fastq.gz</a>
<a class="sourceLine" id="cb3-10" data-line-number="10">[<span class="ex">FLASH</span>]     ./S1.notCombined_1.fastq.gz</a>
<a class="sourceLine" id="cb3-11" data-line-number="11">[<span class="ex">FLASH</span>]     ./S1.notCombined_2.fastq.gz</a>
<a class="sourceLine" id="cb3-12" data-line-number="12">[<span class="ex">FLASH</span>]     ./S1.hist</a>
<a class="sourceLine" id="cb3-13" data-line-number="13">[<span class="ex">FLASH</span>]     ./S1.histogram</a>
<a class="sourceLine" id="cb3-14" data-line-number="14">[<span class="ex">FLASH</span>]</a>
<a class="sourceLine" id="cb3-15" data-line-number="15">[<span class="ex">FLASH</span>] Parameters:</a>
<a class="sourceLine" id="cb3-16" data-line-number="16">[<span class="ex">FLASH</span>]     Min overlap:           10</a>
<a class="sourceLine" id="cb3-17" data-line-number="17">[<span class="ex">FLASH</span>]     Max overlap:           100</a>
<a class="sourceLine" id="cb3-18" data-line-number="18">[<span class="ex">FLASH</span>]     Max mismatch density:  0.250000</a>
<a class="sourceLine" id="cb3-19" data-line-number="19">[<span class="ex">FLASH</span>]     Allow <span class="st">&quot;outie&quot;</span> pairs:   false</a>
<a class="sourceLine" id="cb3-20" data-line-number="20">[<span class="ex">FLASH</span>]     Cap mismatch quals:    false</a>
<a class="sourceLine" id="cb3-21" data-line-number="21">[<span class="ex">FLASH</span>]     Combiner threads:      12</a>
<a class="sourceLine" id="cb3-22" data-line-number="22">[<span class="ex">FLASH</span>]     Input format:          FASTQ, phred_offset=33</a>
<a class="sourceLine" id="cb3-23" data-line-number="23">[<span class="ex">FLASH</span>]     Output format:         FASTQ, phred_offset=33, gzip</a>
<a class="sourceLine" id="cb3-24" data-line-number="24">[<span class="ex">FLASH</span>]</a></code></pre></div>
<p>This has created a series of files named like <code>S${i}.extendedFrags.fastq.gz</code> in which the <code>${i}</code> is replaced by a number between 1 and 384. These files are gzipped fastq files that hold the flashed reads.</p>
</div>
<div id="mappingaligning-the-reads" class="section level3">
<h3>Mapping/Aligning the Reads</h3>
<p>The reference sequences we will use to align to are in a fasta file called <code>gtseq15_loci.fasta</code>. We put that in the <code>map</code> directory and then commence mapping the flashed reads as follows:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># index the reference. Output will be prefixed with &quot;gtseq15_loci&quot;</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="ex">/map/--%</span> bwa index -p gtseq15_loci -a is gtseq15_loci.fasta</a>
<a class="sourceLine" id="cb4-3" data-line-number="3"></a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="co"># here is what comes by on stdout </span></a>
<a class="sourceLine" id="cb4-5" data-line-number="5">[<span class="ex">bwa_index</span>] Pack FASTA... 0.00 sec</a>
<a class="sourceLine" id="cb4-6" data-line-number="6">[<span class="ex">bwa_index</span>] Construct BWT for the packed sequence...</a>
<a class="sourceLine" id="cb4-7" data-line-number="7">[<span class="ex">bwa_index</span>] 0.00 seconds elapse.</a>
<a class="sourceLine" id="cb4-8" data-line-number="8">[<span class="ex">bwa_index</span>] Update BWT... 0.00 sec</a>
<a class="sourceLine" id="cb4-9" data-line-number="9">[<span class="ex">bwa_index</span>] Pack forward-only FASTA... 0.00 sec</a>
<a class="sourceLine" id="cb4-10" data-line-number="10">[<span class="ex">bwa_index</span>] Construct SA from BWT and Occ... 0.00 sec</a>
<a class="sourceLine" id="cb4-11" data-line-number="11">[<span class="ex">main</span>] Version: 0.7.5a-r405</a>
<a class="sourceLine" id="cb4-12" data-line-number="12">[<span class="ex">main</span>] CMD: bwa index -p gtseq15_loci -a is gtseq15_loci.fasta</a>
<a class="sourceLine" id="cb4-13" data-line-number="13">[<span class="ex">main</span>] Real time: 0.061 sec<span class="kw">;</span> <span class="ex">CPU</span>: 0.007 sec</a>
<a class="sourceLine" id="cb4-14" data-line-number="14"></a>
<a class="sourceLine" id="cb4-15" data-line-number="15"></a>
<a class="sourceLine" id="cb4-16" data-line-number="16"><span class="co"># map reads. for each input file, the output is named satro_flashed_s${i}_aln.sam</span></a>
<a class="sourceLine" id="cb4-17" data-line-number="17"><span class="co"># Note the -R to insert appropriate read group header lines</span></a>
<a class="sourceLine" id="cb4-18" data-line-number="18"><span class="ex">/map/--%</span> for i in <span class="dt">{1..384}</span><span class="kw">;</span> <span class="kw">do</span> <span class="ex">bwa</span> mem -aM -v 3 -t 10 -R <span class="st">&quot;@RG\tID:s</span><span class="va">${i}</span><span class="st">\tLB:amplicon\tPL:ILLUMINA\tSM:satro</span><span class="va">${i}</span><span class="st">&quot;</span> ./gtseq15_loci ../flash/S<span class="va">${i}</span>.extendedFrags.fastq.gz  <span class="op">&gt;</span> ./satro_flashed_s<span class="va">${i}</span>_aln.sam<span class="kw">;</span> <span class="kw">done</span></a>
<a class="sourceLine" id="cb4-19" data-line-number="19"></a>
<a class="sourceLine" id="cb4-20" data-line-number="20"></a>
<a class="sourceLine" id="cb4-21" data-line-number="21"><span class="co"># convert the SAMs to BAMs </span></a>
<a class="sourceLine" id="cb4-22" data-line-number="22"><span class="ex">/map/--%</span> for i in <span class="dt">{1..384}</span><span class="kw">;</span> <span class="kw">do</span> <span class="ex">samtools</span> view -bhS satro_flashed_s<span class="va">${i}</span>_aln.sam <span class="op">&gt;</span> satro_flashed_s<span class="va">${i}</span>.bam<span class="kw">;</span> <span class="kw">done</span></a>
<a class="sourceLine" id="cb4-23" data-line-number="23"></a>
<a class="sourceLine" id="cb4-24" data-line-number="24"></a>
<a class="sourceLine" id="cb4-25" data-line-number="25"><span class="co"># Sort the BAMs</span></a>
<a class="sourceLine" id="cb4-26" data-line-number="26"><span class="ex">/map/--%</span> for i in <span class="dt">{1..384}</span><span class="kw">;</span> <span class="kw">do</span> <span class="ex">samtools</span> sort satro_flashed_s<span class="va">${i}</span>.bam -o satro_flashed_s<span class="va">${i}</span>_sorted.bam<span class="kw">;</span> <span class="kw">done</span></a>
<a class="sourceLine" id="cb4-27" data-line-number="27"></a>
<a class="sourceLine" id="cb4-28" data-line-number="28"></a>
<a class="sourceLine" id="cb4-29" data-line-number="29"><span class="co"># Index the BAMs</span></a>
<a class="sourceLine" id="cb4-30" data-line-number="30"><span class="ex">/map/--%</span> for i in <span class="dt">{1..384}</span><span class="kw">;</span> <span class="kw">do</span> <span class="ex">samtools</span> index satro_flashed_s<span class="va">${i}</span>_sorted.bam<span class="kw">;</span> <span class="kw">done</span></a>
<a class="sourceLine" id="cb4-31" data-line-number="31"></a>
<a class="sourceLine" id="cb4-32" data-line-number="32"></a>
<a class="sourceLine" id="cb4-33" data-line-number="33"><span class="co"># Make a text-file list of the BAMs</span></a>
<a class="sourceLine" id="cb4-34" data-line-number="34"><span class="ex">/map/--%</span> ls -1 *sorted.bam <span class="op">&gt;</span> bamlist.txt</a>
<a class="sourceLine" id="cb4-35" data-line-number="35"></a>
<a class="sourceLine" id="cb4-36" data-line-number="36"></a>
<a class="sourceLine" id="cb4-37" data-line-number="37"><span class="co"># generate idx stats and put them in a directory called gtseq17_idx</span></a>
<a class="sourceLine" id="cb4-38" data-line-number="38"><span class="ex">/map/--%</span> mkdir gtseq17_idx</a>
<a class="sourceLine" id="cb4-39" data-line-number="39"><span class="ex">/map/--%</span> for i in <span class="dt">{1..384}</span><span class="kw">;</span> <span class="kw">do</span> <span class="ex">samtools</span> idxstats satro_flashed_s<span class="va">${i}</span>_sorted.bam <span class="op">&gt;</span> ./gtseq17_idx/s<span class="va">${i}</span>_idxstats.txt<span class="kw">;</span> <span class="kw">done</span></a></code></pre></div>
</div>
<div id="variant-detection" class="section level3">
<h3>Variant detection</h3>
<p>We are now set up to detect variants using freebayes. We show how we do that here, but emphasize that if you have already chosen a set of variants that form your microhaplotypes, you could just specify those with a pre-existing VCF file. You <em>don’t</em> need to do new variant detection if you don’t want to. (Of course, if it doesn’t take too long, you may as well do variant detection with every individual you have sequenced at the amplicons, so as to get as many variants as possible.)</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode sh"><code class="sourceCode bash"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co"># index fasta reference file</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="ex">/map/--%</span> samtools faidx gtseq15_loci.fasta</a>
<a class="sourceLine" id="cb5-3" data-line-number="3"> </a>
<a class="sourceLine" id="cb5-4" data-line-number="4"> </a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="co"># call variants, instructing freebayes to NOT return MNPs or other complex variants</span></a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="fu">nohup</span> freebayes-parallel <span class="op">&lt;(</span><span class="ex">fasta_generate_regions.py</span> gtseq15_loci.fasta.fai 150<span class="op">)</span> 10 -f gtseq15_loci.fasta -L bamlist.txt --haplotype-length 0 -kwVa --no-mnps --no-complex <span class="op">&gt;</span> satro384_noMNP_noComplex_noPriors.vcf</a></code></pre></div>
</div>
<div id="runhaplot" class="section level3">
<h3>runHaplot</h3>
<p>After the above is done, the file <code>satro384_noMNP_noComplex_noPriors.vcf</code> can be filtered, if desired, to make sure that everything in it has a solid SNP. Then that file is used with the function <code>prepHaplotFiles</code> to extract haplotypes from the aligned reads in the <code>map</code> directory.<br />
An example of an vcf file and SAM files extracted from an actual GT-seq rockfish data is available in the <code>inst/extdata</code>.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">library</span>(microhaplot)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"></a>
<a class="sourceLine" id="cb6-3" data-line-number="3">run.label &lt;-<span class="st"> &quot;sebastes&quot;</span></a>
<a class="sourceLine" id="cb6-4" data-line-number="4"></a>
<a class="sourceLine" id="cb6-5" data-line-number="5">sam.path &lt;-<span class="st"> </span><span class="kw">tempdir</span>()</a>
<a class="sourceLine" id="cb6-6" data-line-number="6"><span class="kw">untar</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>,</a>
<a class="sourceLine" id="cb6-7" data-line-number="7">                  <span class="st">&quot;sebastes_sam.tar.gz&quot;</span>,</a>
<a class="sourceLine" id="cb6-8" data-line-number="8">                  <span class="dt">package=</span><span class="st">&quot;microhaplot&quot;</span>),</a>
<a class="sourceLine" id="cb6-9" data-line-number="9">      <span class="dt">exdir =</span> sam.path)</a>
<a class="sourceLine" id="cb6-10" data-line-number="10"></a>
<a class="sourceLine" id="cb6-11" data-line-number="11"></a>
<a class="sourceLine" id="cb6-12" data-line-number="12">label.path &lt;-<span class="st"> </span><span class="kw">file.path</span>(sam.path, <span class="st">&quot;label.txt&quot;</span>)</a>
<a class="sourceLine" id="cb6-13" data-line-number="13">vcf.path &lt;-<span class="st"> </span><span class="kw">file.path</span>(sam.path, <span class="st">&quot;sebastes.vcf&quot;</span>)</a>
<a class="sourceLine" id="cb6-14" data-line-number="14"></a>
<a class="sourceLine" id="cb6-15" data-line-number="15"><span class="kw">mvShinyHaplot</span>(<span class="kw">tempdir</span>())</a>
<a class="sourceLine" id="cb6-16" data-line-number="16">app.path &lt;-<span class="st"> </span><span class="kw">file.path</span>(<span class="kw">tempdir</span>(), <span class="st">&quot;microhaplot&quot;</span>)</a>
<a class="sourceLine" id="cb6-17" data-line-number="17"></a>
<a class="sourceLine" id="cb6-18" data-line-number="18">haplo.read.tbl &lt;-<span class="st"> </span><span class="kw">prepHaplotFiles</span>(<span class="dt">run.label =</span> run.label,</a>
<a class="sourceLine" id="cb6-19" data-line-number="19">                            <span class="dt">sam.path =</span> sam.path,</a>
<a class="sourceLine" id="cb6-20" data-line-number="20">                            <span class="dt">out.path =</span> <span class="kw">tempdir</span>(),</a>
<a class="sourceLine" id="cb6-21" data-line-number="21">                            <span class="dt">label.path =</span> label.path,</a>
<a class="sourceLine" id="cb6-22" data-line-number="22">                            <span class="dt">vcf.path =</span> vcf.path,</a>
<a class="sourceLine" id="cb6-23" data-line-number="23">                            <span class="dt">app.path =</span> app.path)</a></code></pre></div>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
